### DISCUSSION TITLE
BusinessRuleAsync ExecuteAsync method error Target object must implement IManageProperties

### DISCUSSION THREAD

**Discussion by jhuerta3** [MarimerLLC/csla#3236] [Questions] [answered]
Hello,

I was hoping someone can give me some guidance. I tried to do some research but I could not find something that I understood. What I notice is if I DONT make the database call in the business rule via the LookUpTableFactory everything works but once I make the call I get the error. This is my first business rule and Im not sure how to solve the implement IManageProperties issue. All my methods are async and my business class inherits from BusinessRuleAsync. 
FYI - I also tried to have sync methods but I got the following error “rule://appsystem.shared.checkstatuschange/StatusID:Synchronous data access is not supported on this runtime. Please use async data access methods"

The error that I get is .

The error: `"Target object must implement IManageProperties"`
The line: `var StatusIDValue = (Guid)ReadProperty(context.Target, PrimaryProperty);`
![HOLD_1](https://user-images.githubusercontent.com/56940135/208821292-fe6c49e3-82cb-4b08-8ec8-cbd9ec5b07fd.png)

Bellow is the business rule. 

```c#
namespace AppSystem.Shared
{
    using Csla.Rules;

    public class CheckStatusChange : BusinessRuleAsync
    {
        Csla.IDataPortal<AppSystem.Shared.LookUpTableInfo> LookUpTableInfoPortal;
        public bool ActiveFlagStatus { get; set; }

        public CheckStatusChange(Csla.Core.IPropertyInfo primaryProperty,
                                 bool activeflagstatus,
                                 Csla.IDataPortal<AppSystem.Shared.LookUpTableInfo> lookuttableinfoportal)
          : base(primaryProperty)
        {
            ActiveFlagStatus = activeflagstatus;
            LookUpTableInfoPortal = lookuttableinfoportal;
        }

        protected override async Task ExecuteAsync(IRuleContext context)
        {
            var StatusIDValue = (Guid)ReadProperty(context.Target, PrimaryProperty);

            if (StatusIDValue.ToString() != null && StatusIDValue.ToString() != "00000000-0000-0000-0000-000000000000")
            {
                LookUpTableFactory objLookFactory = new LookUpTableFactory(LookUpTableInfoPortal);
                Task<LookUpTableInfo> objStatusTask = objLookFactory.GetkLookUpInfoAsync(StatusIDValue);
                LookUpTableInfo objStatus = await objStatusTask;

                if ((objStatus.LookUpName.Equals("Obsolete".ToLower()) || objStatus.LookUpName.Equals("Damaged".ToLower())) && ActiveFlagStatus == true)
                    context.AddErrorResult("Check avilability. Active flag should be false for status Obsolete and Damaged.");
            }
            return;
        }
    }
}
```

**Comment by rockfordlhotka** [comment]
Hi @jhuerta3 - I updated your markdown to make the post more readable - view the markdown to see what I did.

**Comment by rockfordlhotka** [comment, marked as answer]
By default an async rule doesn't populate the `Target` property. This is because an async rule is ... well ... async - and so might be running on another thread, so directly accessing the `Target` object is dangerous due to possible multithreading sync issues.

If you _really do_ want to interact with `Target`, in your rule's constructor you need to set a flag indicating that you understand the danger, and want `Target` to be populated. By doing so, you are accepting that you understand that the object might be manipulated by multiple threads at the same time, and you understand the consequences.

On the other hand, if you just want a property value from the target object, then you should use the `InputProperties` feature - in the rule's constructor list the properties you want to use, so the rules engine can safely load them into the `context`, and then in the `ExecuteAsync` method you can use the collection of input properties provided by the `context` parameter.

**Comment by jhuerta3** [comment]
Hello @rockfordlhotka 

Thank you Sir. I got it working. I changed all my methods to sync and modfied the business rule like so:

        protected override void AddBusinessRules()
        {
            base.AddBusinessRules();
            BusinessRules.AddRule(new CheckStatusChange(StatusIDProperty, StatusNameProperty, ActiveFlagProperty));
        }




    public class CheckStatusChange : BusinessRule
    {
        public CheckStatusChange(Csla.Core.IPropertyInfo StatusIDProperty, 
                                Csla.Core.IPropertyInfo StatusNameProperty, 
                                Csla.Core.IPropertyInfo ActiveFlagProperty)
          : base(StatusIDProperty)
        {
            //Note Order make a difference with InputProperties
            PrimaryProperty = StatusIDProperty;
            InputProperties.Add(StatusIDProperty);
            InputProperties.Add(StatusNameProperty);
            InputProperties.Add(ActiveFlagProperty);
        }

        protected override void Execute(IRuleContext context)
        {
            var statusname = (string)context.InputPropertyValues[InputProperties[1]];
            var activeflag = (bool)context.InputPropertyValues[InputProperties[2]];

            if (!String.IsNullOrEmpty(statusname))
            {
                if ((statusname.ToLower().Equals("Obsolete".ToLower()) || statusname.ToLower().Equals("Damaged".ToLower())) && activeflag == true)
                    context.AddErrorResult("Check avilability. Active flag should be false for status Obsolete and Damaged.");
            }
            return;
        }
    }
